src/Engine/GameObject/Message.hpp

```diff
@@ -3,8 +3,171 @@
+    SpatialTransformTag(const SpatialComponent & spatial) : spatial(spatial) {}
+};
+
+// a spatial's location was set
+struct SpatialPositionSetMessage :  public SpatialTransformTag {
+    SpatialPositionSetMessage(const SpatialComponent & spatial) : SpatialTransformTag(spatial) {}
+};
+
+// a spatial was moved
+struct SpatialMovedMessage :  public SpatialTransformTag {
+    SpatialMovedMessage(const SpatialComponent & spatial) : SpatialTransformTag(spatial) {}
+};
+
+// a spatial's scale was set
+struct SpatialScaleSetMessage :  public SpatialTransformTag {
+    SpatialScaleSetMessage(const SpatialComponent & spatial) : SpatialTransformTag(spatial) {}
+};
+
+// a spatial was scaled
+struct SpatialScaledMessage :  public SpatialTransformTag {
+    SpatialScaledMessage(const SpatialComponent & spatial) : SpatialTransformTag(spatial) {}
+};
+
+// a spatial's rotation was set
+struct SpatialOrientationSetMessage :  public SpatialTransformTag {
+    SpatialOrientationSetMessage(const SpatialComponent & spatial) : SpatialTransformTag(spatial) {}
+};
+
+// a spatial was rotated
+struct SpatialRotatedMessage :  public SpatialTransformTag {
+    SpatialRotatedMessage(const SpatialComponent & spatial) : SpatialTransformTag(spatial) {}
+};
+
+
+
+// a camera was rotated
+struct CameraRotatedMessage : public Message {
+    const CameraComponent & camera;
+    CameraRotatedMessage(const CameraComponent & camera) : camera(camera) {}
+};
+
+
+
+// a collision occurred between the two bounders
+// there may not necessarily have been any adjustment
+// two messages are sent per pair of bounders, where the order of bounders is swapped
+struct CollisionMessage : public Message {
+    const BounderComponent & bounder1, & bounder2;
+    CollisionMessage(const BounderComponent & bounder1, const BounderComponent & bounder2) : bounder1(bounder1), bounder2(bounder2) {}
 };
 
+// a collision occured where norm is the normal of the surface collided with
+// mainly for physics
+struct CollisionNormMessage : public Message {
+    const BounderComponent & bounder;
+    glm::vec3 norm; // normalized
+    CollisionNormMessage(const BounderComponent & bounder, const glm::vec3 & norm) : bounder(bounder), norm(norm) {}
+};
+
+// a collision occurred and the game object's position was adjusted
+struct CollisionAdjustMessage : public Message {
+    const GameObject & gameObject;
+    glm::vec3 delta;
+    CollisionAdjustMessage(const GameObject & gameObject, const glm::vec3 & delta) : gameObject(gameObject), delta(delta) {}
+};
+
+
+
+// the window was resized
+struct WindowSizeMessage : public Message {
+    int width, height;
+    WindowSizeMessage(int width, int height) : width(width), height(height) {}
+};
+
+
+
+// the near/far planes were changed
+struct NearFarMessage : public Message {
+    float near, far;
+    NearFarMessage(float near, float far) : near(near), far(far) {}
+};
+
+
+
+// key input
+struct KeyMessage : public Message {
```

Comment:
Why use this instead of Keyboard::isKeyDown() or Mouse::isButtonDown()

Reply:
it allows for basic event-driven functionality, rather than the need for some sort of update loop. See main. For example, ctrl-g flips gravity. Just a little something I wanted to test and this makes that real simple.